## How this was tested locally

Three layers, in decreasing order of how much I lean on them as evidence: a deterministic unit test (primary), the full gateway regression suite (no regression), and an end-to-end fleet reproduction (supporting, with an honest limit noted at the bottom).

### Layer 1 — deterministic mechanism unit test (primary evidence)

`bcos-gateway/test/unittests/FIB186_BulkDisconnectReactorTest.cpp` is a red→green test that pins the exact invariant the fix must satisfy: **a session-teardown flood must not starve inbound message delivery**, because both used to run on the same unbounded `Host::m_asyncGroup` reactor.

What it does, and why each piece is there:

- `tbb::global_control(max_allowed_parallelism, 2)` pins the shared TBB pool to width 2, so "every reactor worker is occupied by teardown" is deterministic instead of depending on the host core count.
- It builds 8 real `Session` objects (more than the pinned worker count) over a fake ASIO/socket, and sets each session's `messageHandler` to a busy-hold blocker that occupies a reactor worker until a latch is released — this models the per-disconnect work a real bulk-disconnect drives (`onDisconnect -> onRemoveNodeIDs -> syncLatestNodeIDList`).
- It calls the **real** `Session::drop(TCPError)` on all 8, so the teardown notifications take the actual production routing (`Session.cpp` drop path), not a stand-in.
- Once teardown occupies the reactor, it submits a validator-style delivery task through `Host::asyncTo` (the same call `Session::onMessage` uses) and asserts the task runs inside a 50 ms grace window.
- A drain barrier (`done == floodCount + 1`) releases the blockers and joins every task, proving the assertion measured starvation, not a deadlock.

Because it drives the real `drop()` path, the test discriminates between fixes: moving teardown onto a dedicated executor turns it green, while merely throttling teardown on the same reactor would not.

**Red (pre-fix routing).** Reverting only the one-line routing change in `Session::drop` (`postTeardown(` back to `asyncTo(`) and rebuilding:

```
# temporarily: m_server.get().asyncTo( ... )  in Session::drop
cmake --build build --target test-bcos-gateway -j
cd bcos-gateway/test/unittests && ./../../../build/bcos-gateway/test/test-bcos-gateway \
    --run_test=FIB186_BulkDisconnectReactorTest --report_level=short
```

Output:

```
FIB186_BulkDisconnectReactorTest.cpp:214: error: in
"FIB186_BulkDisconnectReactorTest/TeardownFloodMustNotStarveMessageDelivery":
FIB-186 vector D: a PBFT message delivery submitted during a session-teardown flood must not be
starved. ... teardown occupies every worker and delivery never runs, halting consensus.

Test module "Master Test Suite" has failed with:
  1 test case out of 102 failed
  1 assertion out of 2 failed
# exit=201
```

**Green (this PR).** With teardown on `postTeardown` (dedicated `p2pTeardown` executor):

```
cmake --build build --target test-bcos-gateway -j
cd bcos-gateway/test/unittests && ./../../../build/bcos-gateway/test/test-bcos-gateway \
    --run_test=FIB186_BulkDisconnectReactorTest
# *** No errors detected   (time elapsed: 30 ms), exit=0
```

### Layer 2 — full gateway regression suite

The whole `test-bcos-gateway` suite, run from `bcos-gateway/test/unittests` (the config-parsing cases read `data/config/*.ini` relative to cwd):

```
cd bcos-gateway/test/unittests
./../../../build/bcos-gateway/test/test-bcos-gateway --report_level=short
```

```
Test module "Master Test Suite" has passed with:
  102 test cases out of 102 passed
  4689 assertions out of 4689 passed
# exit=0
```

### Layer 3 — end-to-end fleet reproduction (supporting)

A fresh 4-node AIR chain built with the standard tool:

```
tools/BcosAirBuilder/build_chain.sh -p 30300,20200 -l 127.0.0.1:4 -o nodes-churn-test/fresh4 -e <fisco-bcos-binary>
bash nodes-churn-test/fresh4/start_all.sh   # confirm connected=3, view stable, height advancing
```

The gateway derives a peer's P2P id from the RSA modulus of its TLS cert, so each **distinct** client cert signed by the chain's agency CA presents as a **distinct observer session**. Mint N such certs (this is what populates `injcerts/`):

```
CA=nodes-churn-test/fresh4/ca      # ca.crt + ca.key from build_chain
for i in $(seq 1 150); do
  openssl genrsa -out injcerts/c$i.key 2048
  openssl req -new -key injcerts/c$i.key -out injcerts/c$i.csr \
      -subj "/CN=FISCO-BCOS-inj$i/O=fisco-bcos/OU=agency"
  openssl x509 -req -in injcerts/c$i.csr -CA $CA/ca.crt -CAkey $CA/ca.key \
      -CAcreateserial -days 365 -out injcerts/c$i.crt
done
```

The injector establishes all N sessions, holds them, bulk-closes them at once (the FIN/RST storm that drives the server-side `drop()` flood on already-established sessions), then repeats — this is precisely vector D, "persistent bulk-disconnect of established sessions", as opposed to the connect/close churn (vectors A/B) that #5320's admission control already handles.

```python
#!/usr/bin/env python3
"""FIB-186 vector D injector: establish N real gateway sessions (each with a distinct CA-signed
RSA cert -> distinct p2pID) then bulk-disconnect them, repeat."""
import socket, ssl, sys, time, glob, os

CERTD = os.environ.get("INJ_CERTD", "injcerts")

def make_ctx(cert, key):
    ctx = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE          # attacker does not verify the server CN
    ctx.load_cert_chain(certfile=cert, keyfile=key)
    try:
        ctx.set_ciphers("DEFAULT:@SECLEVEL=1")  # allow the fleet's RSA/TLS1.2 ciphers
    except ssl.SSLError:
        pass
    return ctx

def main():
    host = sys.argv[1] if len(sys.argv) > 1 else "127.0.0.1"
    port = int(sys.argv[2]) if len(sys.argv) > 2 else 30300
    cycles = int(sys.argv[3]) if len(sys.argv) > 3 else 20
    hold = float(sys.argv[4]) if len(sys.argv) > 4 else 1.0

    pairs = []
    for crt in sorted(glob.glob(os.path.join(CERTD, "*.crt"))):
        key = crt[:-4] + ".key"
        if os.path.exists(key):
            pairs.append((crt, key))
    ctxs = [make_ctx(c, k) for c, k in pairs]
    print(f"[inj] {len(ctxs)} certs, target {host}:{port}, {cycles} cycles, hold {hold}s", flush=True)

    for cyc in range(cycles):
        conns = []
        ok = 0
        for ctx in ctxs:
            try:
                raw = socket.create_connection((host, port), timeout=3)
                conns.append(ctx.wrap_socket(raw, server_hostname=None))
                ok += 1
            except Exception as e:
                if cyc == 0 and ok == 0:
                    print(f"[inj] handshake failed: {type(e).__name__}: {e}", flush=True)
        print(f"[inj] cycle {cyc}: established {ok}/{len(ctxs)}", flush=True)
        time.sleep(hold)
        for s in conns:            # bulk-disconnect -> server drop() flood
            try: s.close()
            except Exception: pass

if __name__ == "__main__":
    main()
```

```
INJ_CERTD=injcerts python3 dinjector.py 127.0.0.1 30300 40 1.0
```

**Honest limit of layer 3.** On loopback this reproduces the gateway-level trigger — ~140 observer sessions establish and the bulk-close drives the `onRemoveNodeIDs` teardown flood at ~40/s — but it does **not** drive the full permanent consensus halt locally: at that scale both the pre-fix and fixed binaries stayed consensus-healthy. Driving the reactor to the point of a 2f+1 quorum collapse needs botnet-scale simultaneous drops, which a single loopback injector cannot reach — it is bounded by the #5320 accept-rate limiter (100/s) and the process FD limit, and an idle FISCO 3.x chain seals no empty blocks so a load generator is needed to even observe the block-signal stall. This is why the deterministic unit test in layer 1, not the fleet run, is the primary evidence: it isolates the starvation invariant without needing to physically overwhelm the pool.

Environment: macOS arm64, `RelWithDebInfo`, HEAD `6a108994d` (2a executor isolation + 2c node-list-sync debounce). The red run above was produced this session by reverting only the drop-routing line, rebuilding, and running; the tree was restored to `6a108994d` afterwards (`git diff` clean).

--- a/bcos-gateway/bcos-gateway/GatewayFactory.cpp
+++ b/bcos-gateway/bcos-gateway/GatewayFactory.cpp
@@ -788,17 +788,35 @@ std::shared_ptr<Gateway> GatewayFactory::buildGateway(GatewayConfig::Ptr _config
         auto gatewayNodeManagerWeakPtr = std::weak_ptr<GatewayNodeManager>(gatewayNodeManager);
         // register disconnect handler
         service->registerDisconnectHandler(
-            [gatewayNodeManagerWeakPtr](NetworkException e, P2PSession::Ptr p2pSession) {
+            [gatewayNodeManagerWeakPtr, serviceWeakPtr = std::weak_ptr<Service>(service)](
+                NetworkException e, P2PSession::Ptr p2pSession) {
                 if (e.errorCode() == P2PExceptionType::DuplicateSession ||
                     e.errorCode() == P2PExceptionType::Success)
                 {
                     return;
                 }
                 auto gatewayNodeManager = gatewayNodeManagerWeakPtr.lock();
-                if (gatewayNodeManager && p2pSession)
+                if (!gatewayNodeManager || !p2pSession)
                 {
-                    gatewayNodeManager->onRemoveNodeIDs(p2pSession->p2pID());
+                    return;
                 }
+                // FIB-186 (vector D): the teardown notification that drives this handler can be
+                // delayed on the dedicated teardown executor while the SAME peer reconnects -- a
+                // new session for the same p2pID is inserted and its status re-populates the
+                // routing table. removeP2PID keys only on p2pID, so a stale teardown would erase
+                // the new session's routing entries and blank the unicast route until the next
+                // status broadcast. Only remove when this dropped session is still the session of
+                // record for its p2pID (or none is): if a different, current session already owns
+                // the p2pID the peer has reconnected and its entries must be kept.
+                if (auto service = serviceWeakPtr.lock())
+                {
+                    auto current = service->getP2PSessionByNodeId(p2pSession->p2pID());
+                    if (current && current != p2pSession)
+                    {
+                        return;
+                    }
+                }
+                gatewayNodeManager->onRemoveNodeIDs(p2pSession->p2pID());
             });
 
         service->registerUnreachableHandler(

--- a/bcos-gateway/bcos-gateway/gateway/GatewayNodeManager.cpp
+++ b/bcos-gateway/bcos-gateway/gateway/GatewayNodeManager.cpp
@@ -99,7 +99,21 @@ GatewayNodeManager::GatewayNodeManager(std::string const& _uuid, P2pID const& _n
     m_timer->registerTimeoutHandler([this]() {
         if (m_nodeIDListDirty.exchange(false, std::memory_order_acq_rel))
         {
-            syncLatestNodeIDList();
+            // FIB-186 (vector D): syncLatestNodeIDList runs on the timer thread, before
+            // broadcastStatusSeq() re-arms the timer (m_timer->restart() is its first line).
+            // Timer's async_wait handler only logs a throwing timeout handler and does NOT
+            // reschedule, so an exception escaping here would leave the timer un-rearmed and the
+            // seqSync timer would die permanently (all status-seq broadcasts and node-list syncs
+            // would stop). Contain it so broadcastStatusSeq() always runs and re-arms the timer.
+            try
+            {
+                syncLatestNodeIDList();
+            }
+            catch (std::exception const& e)
+            {
+                NODE_MANAGER_LOG(WARNING)
+                    << LOG_DESC("syncLatestNodeIDList exception") << LOG_KV("error", e.what());
+            }
         }
         broadcastStatusSeq();
     });
@@ -320,11 +334,28 @@ void GatewayNodeManager::onRemoveNodeIDs(const P2pID& _p2pID)
         m_p2pID2Seq.erase(_p2pID);
     }
     m_peersRouterTable->removeP2PID(_p2pID);
-    // FIB-186 (vector D): mark the node list dirty instead of syncing inline. A persistent
-    // bulk-disconnect would otherwise drive one full syncLatestNodeIDList (a node-list broadcast to
-    // every front) per dropped session on the teardown executor; the seqSync timer coalesces the
-    // burst into at most one sync per SEQ_SYNC_PERIOD.
-    m_nodeIDListDirty.store(true, std::memory_order_relaxed);
+    // FIB-186 (vector D): notify the local front promptly without re-introducing the per-disconnect
+    // sync flood. removeP2PID above updates the gateway's own routing table immediately; the
+    // front's cached node-list is refreshed by syncLatestNodeIDList. Sync inline only on the first
+    // drop of a burst (the clean->dirty transition); further drops within the window leave the flag
+    // dirty and are coalesced by the seqSync timer (at most one inline sync per SEQ_SYNC_PERIOD, so
+    // a persistent bulk-disconnect cannot flood the front). Without the inline sync the front would
+    // keep a stale node-list for up to a full period, issuing messages to the departed node that
+    // the gateway then drops. This runs on the teardown executor inside Service::onMessage's
+    // try/catch, but wrap it so a throw cannot skip the rest of onDisconnect (other disconnect
+    // handlers, session erase).
+    if (!m_nodeIDListDirty.exchange(true, std::memory_order_acq_rel))
+    {
+        try
+        {
+            syncLatestNodeIDList();
+        }
+        catch (std::exception const& e)
+        {
+            NODE_MANAGER_LOG(WARNING)
+                << LOG_DESC("onRemoveNodeIDs inline sync exception") << LOG_KV("error", e.what());
+        }
+    }
 }
 
 

--- a/bcos-gateway/bcos-gateway/gateway/GatewayNodeManager.h
+++ b/bcos-gateway/bcos-gateway/gateway/GatewayNodeManager.h
@@ -97,10 +97,11 @@ protected:
 
     unsigned const SEQ_SYNC_PERIOD = 1000;
     std::shared_ptr<Timer> m_timer;
-    // FIB-186 (vector D): coalesce disconnect-driven node-list syncs. onRemoveNodeIDs sets this
-    // instead of calling syncLatestNodeIDList() inline; the seqSync timer flushes it at most once
-    // per SEQ_SYNC_PERIOD. Without this a persistent bulk-disconnect drives one full node-list
-    // broadcast to every front per dropped session, piling work onto the teardown executor.
+    // FIB-186 (vector D): coalesce disconnect-driven node-list syncs. onRemoveNodeIDs syncs inline
+    // only on the first drop of a burst (the clean->dirty transition) and otherwise sets this flag;
+    // the seqSync timer flushes it at most once per SEQ_SYNC_PERIOD. So the front is refreshed
+    // promptly on the first drop, while a persistent bulk-disconnect is bounded to ~one sync per
+    // period instead of one full node-list broadcast to every front per dropped session.
     std::atomic_bool m_nodeIDListDirty{false};
 
     GatewayNodeStatusFactory::Ptr m_gatewayNodeStatusFactory;

--- a/bcos-gateway/bcos-gateway/gateway/PeersRouterTable.cpp
+++ b/bcos-gateway/bcos-gateway/gateway/PeersRouterTable.cpp
@@ -105,14 +105,15 @@ void PeersRouterTable::batchInsertNodeList(
         int64_t i = 0;
         for (auto const& nodeID : nodeIDList)
         {
-            if (!m_groupNodeList.count(groupID) || !m_groupNodeList.at(groupID).count(nodeID))
-            {
-                m_groupNodeList[groupID][nodeID] = std::set<P2pID>();
-            }
-            m_groupNodeList[groupID][nodeID].insert(_p2pNodeID);
+            // Capture the map iterators so the reverse index can alias the stable key strings owned
+            // by these nodes (try_emplace leaves an existing entry untouched, inserts otherwise).
+            auto groupIt = m_groupNodeList.try_emplace(groupID).first;
+            auto nodeIt = groupIt->second.try_emplace(nodeID).first;
+            nodeIt->second.insert(_p2pNodeID);
             // FIB-186 (vector D): mirror the insert into the reverse index so the matching removal
-            // is O(K) instead of a full-map scan. std::set dedups repeated (groupID, nodeID) pairs.
-            m_p2pID2GroupNodes[_p2pNodeID].emplace(groupID, nodeID);
+            // is O(K) instead of a full-map scan. Store pointers to the key strings (not copies);
+            // see the lifetime invariant on m_p2pID2GroupNodes. std::set dedups repeated pairs.
+            m_p2pID2GroupNodes[_p2pNodeID].emplace(&groupIt->first, &nodeIt->first);
             if (it->protocol(i))
             {
                 m_nodeProtocolInfo[nodeID] = it->protocol(i);
@@ -151,14 +152,14 @@ void PeersRouterTable::removeP2PIDFromGroupNodeList(const P2pID& _p2pID)
     {
         return;
     }
-    for (auto const& [groupID, nodeID] : revIt->second)
+    for (auto const& [groupIDPtr, nodeIDPtr] : revIt->second)
     {
-        auto groupIt = m_groupNodeList.find(groupID);
+        auto groupIt = m_groupNodeList.find(*groupIDPtr);
         if (groupIt == m_groupNodeList.end())
         {
             continue;
         }
-        auto nodeIt = groupIt->second.find(nodeID);
+        auto nodeIt = groupIt->second.find(*nodeIDPtr);
         if (nodeIt == groupIt->second.end())
         {
             continue;

--- a/bcos-gateway/bcos-gateway/gateway/PeersRouterTable.cpp
+++ b/bcos-gateway/bcos-gateway/gateway/PeersRouterTable.cpp
@@ -110,6 +110,9 @@ void PeersRouterTable::batchInsertNodeList(
                 m_groupNodeList[groupID][nodeID] = std::set<P2pID>();
             }
             m_groupNodeList[groupID][nodeID].insert(_p2pNodeID);
+            // FIB-186 (vector D): mirror the insert into the reverse index so the matching removal
+            // is O(K) instead of a full-map scan. std::set dedups repeated (groupID, nodeID) pairs.
+            m_p2pID2GroupNodes[_p2pNodeID].emplace(groupID, nodeID);
             if (it->protocol(i))
             {
                 m_nodeProtocolInfo[nodeID] = it->protocol(i);
@@ -137,42 +140,40 @@ void PeersRouterTable::removeP2PID(const P2pID& _p2pID)
 void PeersRouterTable::removeP2PIDFromGroupNodeList(const P2pID& _p2pID)
 {
     WriteGuard l(x_groupNodeList);
-    // remove all nodeIDs info belong to p2pID
-    for (auto it = m_groupNodeList.begin(); it != m_groupNodeList.end();)
+    // FIB-186 (vector D): remove only the entries this p2pID actually holds, looked up via the
+    // reverse index, instead of scanning the whole forward map. This bounds the WriteLock hold time
+    // to O(K) (K = entries the peer holds), so a persistent bulk-disconnect on the teardown thread
+    // cannot stall queryP2pIDs (the ReadLock on the unicast routing hot path). Pruning of emptied
+    // node/group entries stays identical to the old full scan: only a set that contained _p2pID can
+    // shrink, so a set the peer never joined is never visited and never pruned.
+    auto revIt = m_p2pID2GroupNodes.find(_p2pID);
+    if (revIt == m_p2pID2GroupNodes.end())
     {
-        for (auto innerIt = it->second.begin(); innerIt != it->second.end();)
+        return;
+    }
+    for (auto const& [groupID, nodeID] : revIt->second)
+    {
+        auto groupIt = m_groupNodeList.find(groupID);
+        if (groupIt == m_groupNodeList.end())
         {
-            for (auto innerIt2 = innerIt->second.begin(); innerIt2 != innerIt->second.end();)
-            {
-                if (*innerIt2 == _p2pID)
-                {
-                    innerIt2 = innerIt->second.erase(innerIt2);
-                }
-                else
-                {
-                    ++innerIt2;
-                }
-            }
-
-            if (innerIt->second.empty())
-            {
-                innerIt = it->second.erase(innerIt);
-            }
-            else
-            {
-                ++innerIt;
-            }
+            continue;
         }
-
-        if (it->second.empty())
+        auto nodeIt = groupIt->second.find(nodeID);
+        if (nodeIt == groupIt->second.end())
         {
-            it = m_groupNodeList.erase(it);
+            continue;
         }
-        else
+        nodeIt->second.erase(_p2pID);
+        if (nodeIt->second.empty())
         {
-            ++it;
+            groupIt->second.erase(nodeIt);
+            if (groupIt->second.empty())
+            {
+                m_groupNodeList.erase(groupIt);
+            }
         }
     }
+    m_p2pID2GroupNodes.erase(revIt);
 }
 
 void PeersRouterTable::updatePeerNodeList(P2pID const& _p2pNodeID, GatewayNodeStatus::Ptr _status)

--- a/bcos-gateway/bcos-gateway/gateway/PeersRouterTable.h
+++ b/bcos-gateway/bcos-gateway/gateway/PeersRouterTable.h
@@ -82,6 +82,15 @@ private:
     // groupID => NodeID => set<P2pID>
     std::map<std::string, std::map<std::string, std::set<P2pID>, std::less<>>, std::less<>>
         m_groupNodeList;
+    // FIB-186 (vector D): reverse index p2pID => the (groupID, nodeID) entries that p2pID appears
+    // under in m_groupNodeList. removeP2PIDFromGroupNodeList used to scan the whole forward map
+    // (O(groups x nodes x p2pIDs)) under x_groupNodeList's WriteLock; under a persistent
+    // bulk-disconnect the teardown thread reacquires that WriteLock once per dropped peer and
+    // blocks queryP2pIDs -- the ReadLock on the unicast routing hot path
+    // (Gateway::asyncSendMessageByNodeID). This index lets removal touch only the entries a peer
+    // actually holds, bounding the WriteLock hold time to O(K). Maintained under x_groupNodeList
+    // together with m_groupNodeList (same lock, so the two are always consistent).
+    std::map<P2pID, std::set<std::pair<std::string, std::string>>> m_p2pID2GroupNodes;
     std::map<std::string, bcos::protocol::ProtocolInfo::ConstPtr> m_nodeProtocolInfo;
     mutable SharedMutex x_groupNodeList;
 

--- a/bcos-gateway/bcos-gateway/gateway/PeersRouterTable.h
+++ b/bcos-gateway/bcos-gateway/gateway/PeersRouterTable.h
@@ -90,7 +90,16 @@ private:
     // (Gateway::asyncSendMessageByNodeID). This index lets removal touch only the entries a peer
     // actually holds, bounding the WriteLock hold time to O(K). Maintained under x_groupNodeList
     // together with m_groupNodeList (same lock, so the two are always consistent).
-    std::map<P2pID, std::set<std::pair<std::string, std::string>>> m_p2pID2GroupNodes;
+    //
+    // Each entry stores POINTERS to the (groupID, nodeID) key strings owned by m_groupNodeList's
+    // nodes, not copies, to avoid duplicating every groupID/nodeID string.
+    // Lifetime invariant: std::map/std::set keep a node's key object at a stable address until that
+    // element is erased, and removeP2PIDFromGroupNodeList erases a (group)/(node) element only when
+    // its p2pID set empties -- i.e. when no peer (including this one) still references it -- so a
+    // pointer stored here always outlives the reverse entry holding it. This REQUIRES
+    // m_groupNodeList to remain a node-based container: do NOT switch it to a flat/vector-backed
+    // map whose keys move on insert/rehash, or these pointers dangle.
+    std::map<P2pID, std::set<std::pair<const std::string*, const std::string*>>> m_p2pID2GroupNodes;
     std::map<std::string, bcos::protocol::ProtocolInfo::ConstPtr> m_nodeProtocolInfo;
     mutable SharedMutex x_groupNodeList;
 

--- a/bcos-gateway/bcos-gateway/libnetwork/Host.cpp
+++ b/bcos-gateway/bcos-gateway/libnetwork/Host.cpp
@@ -807,6 +807,10 @@ void Host::stop()
     {
         m_teardownPool->stop();
     }
+    // A teardown notification that was still running on the pool when it stopped could have posted
+    // follow-up work onto m_asyncGroup after the wait() above returned; drain m_asyncGroup once
+    // more so no such task outlives Host::stop(). wait() on an already-idle group is a cheap no-op.
+    m_asyncGroup.wait();
 }
 bcos::gateway::Host::Host(bcos::crypto::Hash::Ptr _hash,
     std::shared_ptr<ASIOInterface> _asioInterface, std::shared_ptr<SessionFactory> _sessionFactory,

--- a/bcos-gateway/bcos-gateway/libnetwork/Session.cpp
+++ b/bcos-gateway/bcos-gateway/libnetwork/Session.cpp
@@ -369,23 +369,37 @@ void Session::drop(DisconnectReason _reason)
 
     if (m_messageHandler)
     {
-        // FIB-186 (vector D): post the teardown notification onto the dedicated teardown executor,
-        // NOT m_asyncGroup. This handler drives Service::onMessage's error path -> onDisconnect ->
+        // FIB-186 (vector D): run the teardown notification on the dedicated teardown executor, NOT
+        // m_asyncGroup. This handler drives Service::onMessage's error path -> onDisconnect ->
         // onRemoveNodeIDs -> syncLatestNodeIDList; running it on the shared m_asyncGroup let a
         // persistent bulk-disconnect flood starve inter-validator message delivery (which also runs
         // on m_asyncGroup) and permanently halt consensus. postTeardown keeps it off the delivery
         // reactor. Ordering vs message delivery is unchanged: onDisconnect already ran
         // asynchronously and unordered relative to delivery.
-        m_server.get().postTeardown(
-            [self = weak_from_this(), errorCode, errorMsg = std::move(errorMsg)]() {
-                auto session = self.lock();
-                if (!session)
-                {
-                    return;
-                }
-                session->m_messageHandler(
-                    NetworkException(errorCode, errorMsg), session, Message::Ptr());
-            });
+        auto notifyDisconnect = [self = weak_from_this(), errorCode,
+                                    errorMsg = std::move(errorMsg)]() {
+            auto session = self.lock();
+            if (!session)
+            {
+                return;
+            }
+            session->m_messageHandler(
+                NetworkException(errorCode, errorMsg), session, Message::Ptr());
+        };
+        // FIB-186 (vector D): on shutdown Host::stop() has already stopped the teardown executor
+        // (its io_context is stopped and the worker joined), so a postTeardown() here would enqueue
+        // onto a dead pool and the disconnect notification would be silently dropped. Run it inline
+        // instead, mirroring the socket-teardown path below which also switches to inline on
+        // shutdown. haveNetwork() (== Host::m_run) is cleared at the very start of Host::stop(),
+        // before the pool is stopped, so this branch is taken for every drop during shutdown.
+        if (m_server.get().haveNetwork())
+        {
+            m_server.get().postTeardown(std::move(notifyDisconnect));
+        }
+        else
+        {
+            notifyDisconnect();
+        }
     }
 
     // FIB-184: serialize the SSL/socket teardown onto the socket's own (single-threaded)

--- a/bcos-gateway/test/unittests/PeersRouterTableReverseIndexTest.cpp
+++ b/bcos-gateway/test/unittests/PeersRouterTableReverseIndexTest.cpp
@@ -1,0 +1,145 @@
+/**
+ *  Copyright (C) 2026 FISCO BCOS.
+ *  SPDX-License-Identifier: Apache-2.0
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ * @brief Equivalence test for the FIB-186 (vector D) reverse-index optimisation of
+ *        PeersRouterTable::removeP2PIDFromGroupNodeList. The removal was changed from an
+ *        O(groups x nodes x p2pIDs) scan of the whole forward map to an O(K) targeted removal
+ * driven by a reverse index (p2pID -> its (groupID, nodeID) entries), to shorten the
+ * x_groupNodeList WriteLock hold time so a bulk-disconnect flood on the teardown thread cannot
+ * stall queryP2pIDs on the routing hot path. This test pins that the targeted removal produces the
+ *        same observable routing table as the old full scan: the removed peer disappears from every
+ *        (group, node) it held, other peers are untouched, emptied node/group entries are pruned,
+ *        the reverse index is cleared so a re-insert + remove stays correct, and removing an
+ * unknown peer is a no-op.
+ * @file PeersRouterTableReverseIndexTest.cpp
+ * @date 2026-07-16
+ */
+
+#include "bcos-framework/protocol/ProtocolInfo.h"
+#include "bcos-gateway/gateway/PeersRouterTable.h"
+#include "bcos-gateway/protocol/GatewayNodeStatus.h"
+#include "bcos-utilities/testutils/TestPromptFixture.h"
+#include <boost/test/unit_test.hpp>
+#include <set>
+#include <string>
+#include <vector>
+
+using namespace bcos;
+using namespace bcos::gateway;
+using namespace bcos::protocol;
+using namespace bcos::test;
+
+BOOST_FIXTURE_TEST_SUITE(PeersRouterTableReverseIndexTest, TestPromptFixture)
+
+namespace
+{
+// Promote the two protected mutators so the test can drive the forward map and its reverse index in
+// isolation, without the peer-status / gateway-info bookkeeping of the public updatePeerStatus.
+class ExposedPeersRouterTable : public PeersRouterTable
+{
+public:
+    ExposedPeersRouterTable() : PeersRouterTable("", nullptr, nullptr) {}
+    using PeersRouterTable::batchInsertNodeList;
+    using PeersRouterTable::removeP2PIDFromGroupNodeList;
+};
+
+GroupNodeInfo::Ptr makeGroupNodeInfo(std::string const& _group, std::vector<std::string> _nodes)
+{
+    auto info = std::make_shared<bcostars::protocol::GroupNodeInfoImpl>();
+    info->setGroupID(_group);
+    // batchInsertNodeList reads protocol(i) for each node index, so append one protocol per node.
+    for (std::size_t i = 0; i < _nodes.size(); ++i)
+    {
+        info->appendProtocol(std::make_shared<ProtocolInfo>(
+            ProtocolModuleID::NodeService, ProtocolVersion::V1, ProtocolVersion::V1));
+    }
+    info->setNodeIDList(std::move(_nodes));
+    return info;
+}
+
+std::set<P2pID> S(std::initializer_list<std::string> _l)
+{
+    return std::set<P2pID>(_l.begin(), _l.end());
+}
+}  // namespace
+
+// The removed peer disappears from every (group, node) it held; co-located peers are untouched; a
+// (group, node) whose only holder was the removed peer is pruned to empty.
+BOOST_AUTO_TEST_CASE(TargetedRemovalMatchesFullScan)
+{
+    ExposedPeersRouterTable table;
+
+    // p2pA holds group1{n1,n2}, group2{n3}
+    table.batchInsertNodeList(
+        "A", {makeGroupNodeInfo("group1", {"n1", "n2"}), makeGroupNodeInfo("group2", {"n3"})});
+    // p2pB holds group1{n1}, group2{n3,n4}
+    table.batchInsertNodeList(
+        "B", {makeGroupNodeInfo("group1", {"n1"}), makeGroupNodeInfo("group2", {"n3", "n4"})});
+    // p2pC holds group1{n2}
+    table.batchInsertNodeList("C", {makeGroupNodeInfo("group1", {"n2"})});
+    // p2pD is the sole holder of group3{n9} -- removing D must prune the node and the whole group.
+    table.batchInsertNodeList("D", {makeGroupNodeInfo("group3", {"n9"})});
+
+    // Pre-removal routing view.
+    BOOST_CHECK(table.queryP2pIDs("group1", "n1") == S({"A", "B"}));
+    BOOST_CHECK(table.queryP2pIDs("group1", "n2") == S({"A", "C"}));
+    BOOST_CHECK(table.queryP2pIDs("group2", "n3") == S({"A", "B"}));
+    BOOST_CHECK(table.queryP2pIDs("group2", "n4") == S({"B"}));
+    BOOST_CHECK(table.queryP2pIDsByGroupID("group3") == S({"D"}));
+
+    table.removeP2PIDFromGroupNodeList("A");
+
+    // A is gone from every entry it held; B and C are untouched; no entry A did not hold changed.
+    BOOST_CHECK(table.queryP2pIDs("group1", "n1") == S({"B"}));
+    BOOST_CHECK(table.queryP2pIDs("group1", "n2") == S({"C"}));
+    BOOST_CHECK(table.queryP2pIDs("group2", "n3") == S({"B"}));
+    BOOST_CHECK(table.queryP2pIDs("group2", "n4") == S({"B"}));
+
+    table.removeP2PIDFromGroupNodeList("D");
+
+    // D was the sole holder of group3{n9}: the node set empties and the group is pruned away.
+    BOOST_CHECK(table.queryP2pIDs("group3", "n9").empty());
+    BOOST_CHECK(table.queryP2pIDsByGroupID("group3").empty());
+}
+
+// The reverse index is cleared when a peer is removed, so a later re-insert (the remove-then-insert
+// shape of updatePeerStatus) and a second removal only affect the peer's new entries.
+BOOST_AUTO_TEST_CASE(ReverseIndexClearedOnRemovalAndIdempotent)
+{
+    ExposedPeersRouterTable table;
+
+    table.batchInsertNodeList("A", {makeGroupNodeInfo("group1", {"n1"})});
+    table.batchInsertNodeList("B", {makeGroupNodeInfo("group1", {"n1"})});
+    BOOST_CHECK(table.queryP2pIDs("group1", "n1") == S({"A", "B"}));
+
+    // Remove A, then re-insert it under a fresh (group1, n5) only.
+    table.removeP2PIDFromGroupNodeList("A");
+    BOOST_CHECK(table.queryP2pIDs("group1", "n1") == S({"B"}));
+    table.batchInsertNodeList("A", {makeGroupNodeInfo("group1", {"n5"})});
+    BOOST_CHECK(table.queryP2pIDs("group1", "n5") == S({"A"}));
+
+    // Second removal must clear only the new entry (proves the first removal cleared A's old
+    // reverse entries -- a stale (group1, n1) reverse entry would have wrongly evicted B here).
+    table.removeP2PIDFromGroupNodeList("A");
+    BOOST_CHECK(table.queryP2pIDs("group1", "n5").empty());
+    BOOST_CHECK(table.queryP2pIDs("group1", "n1") == S({"B"}));
+
+    // Removing a never-seen peer, and removing A a third time, are no-ops (no crash, B intact).
+    table.removeP2PIDFromGroupNodeList("A");
+    table.removeP2PIDFromGroupNodeList("unknown-p2p");
+    BOOST_CHECK(table.queryP2pIDs("group1", "n1") == S({"B"}));
+}
+
+BOOST_AUTO_TEST_SUITE_END()

--- a/bcos-gateway/test/unittests/PeersRouterTableReverseIndexTest.cpp
+++ b/bcos-gateway/test/unittests/PeersRouterTableReverseIndexTest.cpp
@@ -142,4 +142,39 @@ BOOST_AUTO_TEST_CASE(ReverseIndexClearedOnRemovalAndIdempotent)
     BOOST_CHECK(table.queryP2pIDs("group1", "n1") == S({"B"}));
 }
 
+// The reverse index aliases m_groupNodeList's key strings by pointer. A group/node fully erased and
+// later recreated must not corrupt routing: the recreated node owns fresh key objects, and no stale
+// reverse entry may reference the freed ones. (Without ASan this is a probabilistic stress of the
+// lifetime invariant, not a proof; the invariant itself is argued on m_p2pID2GroupNodes.)
+BOOST_AUTO_TEST_CASE(GroupErasedAndRecreatedKeepsRoutingCorrect)
+{
+    ExposedPeersRouterTable table;
+
+    // P is the sole holder of group9{n1,n2}; removing P erases both nodes and the whole group.
+    table.batchInsertNodeList("P", {makeGroupNodeInfo("group9", {"n1", "n2"})});
+    BOOST_CHECK(table.queryP2pIDs("group9", "n1") == S({"P"}));
+    table.removeP2PIDFromGroupNodeList("P");
+    BOOST_CHECK(table.queryP2pIDsByGroupID("group9").empty());
+
+    // Churn many unrelated groups so the allocator is likely to reuse the memory the erased group9
+    // key strings occupied.
+    for (int i = 0; i < 64; ++i)
+    {
+        table.batchInsertNodeList(
+            "X", {makeGroupNodeInfo("churn" + std::to_string(i), {"a", "b"})});
+    }
+
+    // Recreate group9 through a different peer Q; its keys are fresh objects at (possibly) reused
+    // addresses.
+    table.batchInsertNodeList("Q", {makeGroupNodeInfo("group9", {"n1"})});
+    BOOST_CHECK(table.queryP2pIDs("group9", "n1") == S({"Q"}));
+
+    // Removing X keys off X's own churn entries only; Q's recreated group9 entry is untouched.
+    table.removeP2PIDFromGroupNodeList("X");
+    BOOST_CHECK(table.queryP2pIDs("churn0", "a").empty());
+    BOOST_CHECK(table.queryP2pIDs("group9", "n1") == S({"Q"}));
+    table.removeP2PIDFromGroupNodeList("Q");
+    BOOST_CHECK(table.queryP2pIDsByGroupID("group9").empty());
+}
+
 BOOST_AUTO_TEST_SUITE_END()

